torch/distributed/_sharding_spec/api.py

```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+                of the following remote formats:
+
+                    1. "rank:<rank>/<device>" (ex: "rank:0/cuda:0").
+                    2. "<worker_name>/<device>" (ex: "trainer0/cuda:0").
+        """
+
+        ShardPlacement = Union[Device, PlacementSpec]
+
+        def __init__(
+                self,
+                shard_offsets: List[int],
+                shard_lengths: List[int],
+                placement: ShardPlacement):
+
+            if not isinstance(placement, PlacementSpec) and not is_valid_device(placement):
+                raise ValueError(f'{placement} is not a valid device')
+
+            if len(shard_offsets) != len(shard_lengths):
+                raise ValueError(
+                    f'shard_offsets and shard_lengths should have '
+                    f'the same number of elements, found {len(shard_offsets)} '
+                    f'and {shard_lengths} respectively')
+
+            for i in range(len(shard_offsets)):
+                if shard_offsets[i] < 0:
+                    raise ValueError('shard_offsets should be >=0')
+                if shard_lengths[i] <= 0:
+                    raise ValueError('shard_lengths should be > 0')
+
+            self._shard_offsets = shard_offsets
+            self._shard_lengths = shard_lengths
+            self._placement = placement
+
+        def __repr__(self):
+            return (
+                f'Shard(shard_offsets: {self._shard_offsets}, '
+                f'shard_lengths: {self._shard_lengths}, placement: {self._placement})'
+            )
+
+        @property
+        def shard_offsets(self):
+            return self._shard_offsets
+
+        @property
+        def shard_lengths(self):
+            return self._shard_lengths
+
+        @property
+        def placement(self):
+            return self._placement
+
+    def __init__(self, shards: List[Shard]):
+        """
+        This is a type of PlacementSpec that allows users to specify a generic
+        sharding scheme by specifying exactly how each shard is laid out.
+
+        Args:
+            shards(List[Shard]): List of :class:`Shard` objects representing
+                each shard.
+        """
+        super(GenericShardingSpec, self).__init__()
+        if len(shards) == 0:
+            raise ValueError(f'Empty shard list provided: {shards}')
+
+        # Validate each shard has same rank.
+        rank = -1
+        for shard in shards:
+            if rank != -1 and rank != len(shard.shard_offsets):
+                raise ValueError(f'Found inconsistent ranks for shards: {rank} and {len(shard.shard_offsets)}')
+            rank = len(shard.shard_offsets)
+
+        self._validate_non_overlapping(shards)
+
+        self._shards = shards
+
+    @staticmethod
+    def _validate_non_overlapping(shards: List[Shard]):
+        """
+        Ensures none of the shards overlap with each other.
+        """
+        for i in range(len(shards)):
```

Comment:
This O(N^2) cost seems to be suboptimal. Should be faster if we sort the shards first. Probably add a TODO here to avoid missing the optimization opporunity?